packages/service/src/index.ts

```diff
@@ -3,35 +3,44 @@ import * as fs from 'fs'
 import * as grpc from 'grpc'
 import * as protoLoader from '@grpc/proto-loader'
 import * as path from 'path'
+import Runner from '@mesg/orchestrator/lib/runner'
 import { decode, encode } from '@mesg/api/lib/util/encoder'
+import * as base58 from '@mesg/api/lib/util/base58'
 import { IExecution } from '@mesg/api/lib/execution';
 import { EventCreateOutputs } from '@mesg/api/lib/event';
 import { EventEmitter } from 'events'
 
 type Options = {
   endpoint?: string
-  payload?: string
+  signature?: string
   definition?: any
 }
 
 class Service {
   private definition: any
   private tasks: Tasks
-  private _client: any
-  private credential: Promise<grpc.Metadata>
+  private _client: Promise<any>
+  private _token: grpc.Metadata
 
   constructor(options: Options = {}) {
     this.definition = options.definition || YAML.safeLoad(fs.readFileSync('./mesg.yml').toString());
-    const protoDefinition = grpc.loadPackageDefinition(protoLoader.loadSync(path.join(__dirname, 'runner.proto'))) as any
-    this._client = new protoDefinition.mesg.grpc.runner.Runner(options.endpoint || process.env.MESG_ENDPOINT, grpc.credentials.createInsecure())
-    this.credential = this.register(options.payload || process.env.MESG_REGISTER_PAYLOAD)
+    this._client = this.register(
+      options.signature || process.env.MESG_REGISTER_SIGNATURE,
+      options.endpoint || process.env.MESG_ENDPOINT,
+      process.env.MESG_SERVICE_HASH,
+      process.env.MESG_ENV_HASH
+    )
   }
 
-  async register(payload: string): Promise<grpc.Metadata> {
-    const res = await this.unaryCall('Register', { payload })
-    const meta = new grpc.Metadata()
-    meta.add('mesg_credential_token', res.token)
-    return meta
+  async register(signature: string, endpoint: string, serviceHash: string, envHash: string): Promise<grpc.GrpcObject> {
+    const runner = new Runner(endpoint)
+    const { token } = await runner.register({ serviceHash, envHash }, signature)
+    this._token = new grpc.Metadata()
+    this._token.add('mesg_credential_token', token)
+    const { mesg } = grpc.loadPackageDefinition(protoLoader.loadSync(path.join(__dirname, 'runner', 'runner.proto'), {
+      includeDirs: [__dirname]
+    })) as any
+    return new mesg.grpc.runner.Runner(endpoint, grpc.credentials.createInsecure())
```

Comment:
Isn't possible to pass the token metadata directly to the gRPC client like in Go?
Like this, you don't need to pass the token metadata to each request.

Reply:
You can combine credentials and use a `createFromMetadataGenerator` but unfortunately, there is a runtime error because these kinds of credentials cannot be used with `insecure` connection